cmd/prometheus/main.go

```diff
@@ -94,13 +97,8 @@ func main() {
 
 	a.HelpFlag.Short('h')
 
-	a.Flag("log.level",
-		"Only log messages with the given severity or above. One of: [debug, info, warn, error, fatal]").
-		Default("info").StringVar(&cfg.logLevel)
-
-	a.Flag("log.format",
-		`Set the log target and format. Example: "logger:syslog?appname=bob&local=7" or "logger:stdout?json=true"`).
-		Default("logger:stderr").StringVar(&cfg.logFormat)
+	a.Flag(promlog.LevelFlagName, promlog.LevelFlagHelp).
```

Comment:
> This would require our logging package to depend on our chosen flag package, which is very unnecessary.

That would be the simplest and most obvious way to do it, but it's not the only way.

> In a more general sense, as it keeps coming back as a point we disagree on: not every character of code duplication is bad

I agree in general. There's a little one-off copying, and then there's a little copying that would regularly need to be recopied into tens if not hundreds of exporters. This is the latter class.

This fragment may be trivial now, but what happens as more flags are added?

> Example: we switched to a different flag package in prometheus/prometheus. All other applications should to, but it will propagate gradually.

That's a different class of change, it's an intrusive labour intensive change that's unavoidable if you want to change to a different library.

My problem with your proposal is that for every new feature to the log library, it takes what could be a trivial change for library users and changes it instead into a labour intensive change.

> Abstracting every trivial fragment into some package makes our case bases rigid and changes significantly more painful. At the same time it does not actually do a very good job in making things more unified.

I disagree, having the names and meanings of our flags in a package that exporters can import and call ~1 function for makes changes far easier, and ensures that all exporters can easily keep up to date and in sync.


We already have a standard low-effort way of managing that, and that's vendoring. If adding flags or new output formats to the logging library requires anything beyond updating vendoring, then that's the wrong design as it's creating unnecessary effort for tens to hundreds of developers.